File: validators/cybo_ecosafety_validator.cpp
#include <string>
#include <vector>
#include <stdexcept>
#include <iostream>

// Assume you have a thin JSON wrapper; replace with your actual library.
#include "json.hpp" // nlohmann/json or equivalent

using json = nlohmann::json;

struct ValidationConfig {
    std::string expected_grammar_id;
    std::string owner_did;
    std::vector<std::string> forbidden_labels;  // e.g. {"DigitalWaterTwin", "Digital Water Twin"};
};

static std::string compute_hex_stamp(const json& shard_core);

bool verify_hex_stamp(const json& shard, const ValidationConfig& cfg) {
    if (!shard.contains("hex_stamp")) return false;

    // Define the "core" to be hashed: core_math ∥ grammar_id ∥ did_author
    json core;
    core["grammar_id"]  = shard.value("grammar_id", "");
    core["did_author"]  = shard.value("did_author", "");
    core["ker"]         = shard.value("ker", json::object());
    core["residual"]    = shard.value("residual", json::object());
    core["fields"]      = shard.value("fields", json::array());

    const std::string expected = compute_hex_stamp(core);
    const std::string provided = shard.value("hex_stamp", "");

    return (expected == provided);
}

bool labels_ok(const json& shard, const ValidationConfig& cfg) {
    if (!shard.contains("labels")) return true;
    const auto labels = shard.at("labels");
    if (!labels.is_array()) return false;

    for (const auto& lbl_json : labels) {
        if (!lbl_json.is_string()) continue;
        const std::string lbl = lbl_json.get<std::string>();
        for (const auto& forbidden : cfg.forbidden_labels) {
            if (lbl.find(forbidden) != std::string::npos) {
                return false;
            }
        }
    }
    return true;
}

bool check_origin_and_grammar(const json& shard, const ValidationConfig& cfg) {
    const std::string grammar_id = shard.value("grammar_id", "");
    const std::string origin     = shard.value("origin", "");
    const std::string did_author = shard.value("did_author", "");

    if (grammar_id == cfg.expected_grammar_id) {
        // Shards using your grammar_id must be authored by you and marked internal
        if (did_author != cfg.owner_did) return false;
        if (origin != "internal")        return false;
    }
    return true;
}

bool validate_shard(const json& shard, const ValidationConfig& cfg) {
    if (!check_origin_and_grammar(shard, cfg)) {
        std::cerr << "Origin/grammar/DID mismatch.\n";
        return false;
    }
    if (!labels_ok(shard, cfg)) {
        std::cerr << "Forbidden external branding label detected.\n";
        return false;
    }
    if (!verify_hex_stamp(shard, cfg)) {
        std::cerr << "Hex stamp verification failed.\n";
        return false;
    }
    return true;
}

// ---- placeholder hash implementation ----
// Replace this with your chosen hash (e.g., SHA-256 via OpenSSL/libsodium/etc.).
#include <sstream>
static std::string compute_hex_stamp(const json& shard_core) {
    std::string s = shard_core.dump();
    // Simple non-cryptographic placeholder: FNV-1a 64-bit
    const uint64_t FNV_OFFSET = 1469598103934665603ULL;
    const uint64_t FNV_PRIME  = 1099511628211ULL;
    uint64_t hash = FNV_OFFSET;
    for (unsigned char c : s) {
        hash ^= c;
        hash *= FNV_PRIME;
    }
    std::ostringstream oss;
    oss << std::hex << hash;
    return oss.str();
}

// Simple CLI entrypoint
int main(int argc, char** argv) {
    if (argc < 4) {
        std::cerr << "Usage: cybo_ecosafety_validator <json_file> <expected_grammar_id> <owner_did>\n";
        return 1;
    }

    const std::string json_path         = argv[1];
    const std::string expected_grammar  = argv[2];
    const std::string owner_did         = argv[3];

    ValidationConfig cfg;
    cfg.expected_grammar_id = expected_grammar;
    cfg.owner_did           = owner_did;
    cfg.forbidden_labels    = {"DigitalWaterTwin", "Digital Water Twin", "SmartWaterTwin"};

    std::ifstream in(json_path);
    if (!in) {
        std::cerr << "Cannot open file: " << json_path << "\n";
        return 1;
    }
    json shard;
    in >> shard;

    if (!validate_shard(shard, cfg)) {
        std::cerr << "Shard validation FAILED for " << json_path << "\n";
        return 2;
    }

    std::cout << "Shard validation OK for " << json_path << "\n";
    return 0;
}

File: validators/cybo_ecosafety_validator.rs
use serde::Deserialize;
use serde_json::Value;
use std::fs;

#[derive(Deserialize)]
struct KER {
    knowledge_k: f64,
    eco_impact_e: f64,
    risk_of_harm_r: f64,
}

#[derive(Deserialize)]
struct Residual {
    vt: f64,
    vt_prev: f64,
}

#[derive(Deserialize)]
struct Field {
    param_id: String,
    raw_value: f64,
    unit: String,
    rx_value: f64,
    corridor_id: String,
}

#[derive(Deserialize)]
struct Shard {
    shard_id: String,
    node_id: String,
    grammar_id: String,
    origin: String,
    did_author: String,
    timestamp_utc: String,
    ker: KER,
    residual: Residual,
    fields: Vec<Field>,
    labels: Option<Vec<String>>,
    hex_stamp: String,
}

struct ValidationConfig {
    expected_grammar_id: String,
    owner_did: String,
    forbidden_labels: Vec<String>,
}

fn compute_hex_stamp(core: &Value) -> String {
    use std::hash::{Hash, Hasher};
    use std::collections::hash_map::DefaultHasher;

    let s = core.to_string();
    let mut hasher = DefaultHasher::new();
    s.hash(&mut hasher);
    let hash = hasher.finish();
    format!("{:x}", hash)
}

fn verify_hex_stamp(shard: &Shard, raw: &Value, cfg: &ValidationConfig) -> bool {
    // Build core subset from raw JSON
    let mut core = serde_json::Map::new();
    core.insert("grammar_id".into(), raw["grammar_id"].clone());
    core.insert("did_author".into(), raw["did_author"].clone());
    core.insert("ker".into(), raw["ker"].clone());
    core.insert("residual".into(), raw["residual"].clone());
    core.insert("fields".into(), raw["fields"].clone());

    let core_val = Value::Object(core);
    let expected = compute_hex_stamp(&core_val);
    expected == shard.hex_stamp
}

fn labels_ok(shard: &Shard, cfg: &ValidationConfig) -> bool {
    if let Some(labels) = &shard.labels {
        for lbl in labels {
            for forbidden in &cfg.forbidden_labels {
                if lbl.contains(forbidden) {
                    return false;
                }
            }
        }
    }
    true
}

fn check_origin_and_grammar(shard: &Shard, cfg: &ValidationConfig) -> bool {
    if shard.grammar_id == cfg.expected_grammar_id {
        if shard.did_author != cfg.owner_did {
            return false;
        }
        if shard.origin != "internal" {
            return false;
        }
    }
    true
}

fn validate_shard(shard: &Shard, raw: &Value, cfg: &ValidationConfig) -> bool {
    if !check_origin_and_grammar(shard, cfg) {
        eprintln!("Origin/grammar/DID mismatch.");
        return false;
    }
    if !labels_ok(shard, cfg) {
        eprintln!("Forbidden external branding label detected.");
        return false;
    }
    if !verify_hex_stamp(shard, raw, cfg) {
        eprintln!("Hex stamp verification failed.");
        return false;
    }
    true
}

fn main() {
    let args: Vec<String> = std::env::args().collect();
    if args.len() < 4 {
        eprintln!("Usage: cybo_ecosafety_validator <json_file> <expected_grammar_id> <owner_did>");
        std::process::exit(1);
    }

    let json_path        = &args[1];
    let expected_grammar = &args[2];
    let owner_did        = &args[3];

    let cfg = ValidationConfig {
        expected_grammar_id: expected_grammar.clone(),
        owner_did: owner_did.clone(),
        forbidden_labels: vec![
            "DigitalWaterTwin".into(),
            "Digital Water Twin".into(),
            "SmartWaterTwin".into(),
        ],
    };

    let data = fs::read_to_string(json_path)
        .expect("Cannot read JSON file");
    let raw: Value = serde_json::from_str(&data)
        .expect("Invalid JSON");
    let shard: Shard = serde_json::from_value(raw.clone())
        .expect("JSON does not match Shard schema");

    if !validate_shard(&shard, &raw, &cfg) {
        eprintln!("Shard validation FAILED for {}", json_path);
        std::process::exit(2);
    }

    println!("Shard validation OK for {}", json_path);
}
